pkg/util/clientauthorizer/admin.go

```diff
@@ -5,23 +5,68 @@ package clientauthorizer
 
 import (
 	"crypto/tls"
-
-	"github.com/sirupsen/logrus"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
 )
 
+// NewAdmin creates a new instance of ClientAuthorizer to be used with Admin API.
+// This authorizer allows connections only if they
+// contain a valid client certificate signed by `caBundlePath` and
+// the client certificate's CommonName equals `clientCertCommonName`.
+func NewAdmin(caBundlePath, clientCertCommonName string) (ClientAuthorizer, error) {
+	authorizer := &admin{
+		clientCertCommonName: clientCertCommonName,
+		readFile:             ioutil.ReadFile,
+	}
+
+	return authorizer, authorizer.readCABundle(caBundlePath)
+}
+
 type admin struct {
-	log *logrus.Entry
+	roots                *x509.CertPool
+	clientCertCommonName string
+
+	readFile func(filename string) ([]byte, error)
 }
 
-func NewAdmin(log *logrus.Entry) ClientAuthorizer {
-	return &admin{
-		log: log,
+func (a *admin) readCABundle(caBundlePath string) error {
+	caBundle, err := a.readFile(caBundlePath)
+	if err != nil {
+		return err
+	}
+
+	roots := x509.NewCertPool()
+	ok := roots.AppendCertsFromPEM(caBundle)
+	if !ok {
+		return fmt.Errorf("can not decode admin CA bundle from %s", caBundlePath)
 	}
+
+	a.roots = roots
+	return nil
 }
 
 func (a *admin) IsAuthorized(cs *tls.ConnectionState) bool {
-	a.log.Print("Admin auth is not implemented yet")
-	return false
+	if cs == nil || len(cs.PeerCertificates) == 0 {
+		return false
+	}
+
+	verifyOpts := x509.VerifyOptions{
+		Roots:         a.roots,
+		Intermediates: x509.NewCertPool(),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+
+	for _, cert := range cs.PeerCertificates[1:] {
+		verifyOpts.Intermediates.AddCert(cert)
+	}
+
+	_, err := cs.PeerCertificates[0].Verify(verifyOpts)
+	if err != nil {
+		return false
```

Comment:
Would we ever want to know what this error is?

Reply:
Yeah, I thought about it too. I'm not sure how useful this error is: it will mean that the certificate is not valid. If someone, somehow will be bombarding us with an invalid client certificate - we will have a lot of noise in the logs. On the other after the expiery of valid certificate we will be declining requests to the admin API without logging it.

Reply:
I think, I'll add logging here. The situation when someone will be sending invalid certificate on porpuse is very unlikely because admin API is not exposed to public, but it is useful to have something in logs in case we miss certificate expiration date or something else goes wrong on our side.

Thank you for raising this concern, Nils!